### ISSUE TITLE
Moss integration

### ISSUE THREAD

**Issue by jonatanschroeder** [PrairieLearn/PrairieLearn#3659]
It would be awesome to have moss integration, where all files for an assessment or assessment question with file submissions (pl-file-editor, pl-file-upload) are submitted to moss and the report is provided back.

**Comment by jonatanschroeder** [comment]
An alternative to Moss is Dolos: https://dolos.ugent.be/, https://github.com/dodona-edu/dolos. This has an MIT license and an API for integration, including a [JS library](https://www.npmjs.com/package/@dodona/dolos-lib).

**Comment by mwest1066** [comment]
Oh yes! This would be very nice to integrate.

**Comment by ZacWarham** [comment]
Just another vote for this. Would be interested in cross referencing coding and essay style questions in both pl-file-editor and pl-rich-text-editor. Other use cases would be for notebook/md/pdf files and also for comparing pictures (i.e. student working through pl-image-capture or pl-file-upload)

**Comment by jonatanschroeder** [comment]
I think working with images would be a different methodology. Tools like Moss and Dolos are optimized for content that can be parsed into tokens, which can then be compared using some algorithmic strategy. Comparison between images and PDF files would be trickier to implement using the same tools. Similarly, I'd assume that tools that are optimized for images would likely not work as well with code/text.

Side note: I was actually just playing around with dolos for exploration only, and it has some pretty big dependencies, and installed over 400 MB on a single package, mainly related to support for multiple languages (reported upstream, https://github.com/dodona-edu/dolos/issues/1839), which would significantly increase the size of the PrairieLearn image. So it may be something to consider in the future, but at this point might not be suitable right away.

**Comment by jonatanschroeder** [comment]
Note that moss has just released an NPM package (https://github.com/abdelhalimyasser/node-moss), but it's just a client that connects to moss's own servers.

**Comment by ZacWarham** [comment]
> working with images would be a different methodology

Yes definitely. I suspect this issue would get broken into different components instead of having diverging issues on plagiarism until it is a consideration

**Comment by reteps** [comment]
The MOSS system in particular has issues because students can trivially rewrite code they got from a friend so that it won't pick it up. Is this mainly for essays?

**Comment by jonatanschroeder** [comment]
> The MOSS system in particular has issues because students can trivially rewrite code they got from a friend so that it won't pick it up. Is this mainly for essays?

Both Moss and Dolos are built to catch this kind of change. They're able to identify renaming variables/functions, changes in indentation, changes in comments, and other similar rewordings that are typical of students copying code from each other. They don't "catch" cheating, they provide a similarity value based on a tokenization of the content, not on string comparison (which is why they are optimized for code instead of essays). It is a matter of having staff review those similarities and evaluate those above a threshold to make a decision on if there are signs of potential collaboration.

**Comment by reteps** [comment]
Hmm, ok. I am aware of a few workarounds, but I guess this is catch "90% of cheaters" case.
